working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1227,69 +1347,90 @@ var (a, int b) = ... // Schema is `(?, int)`.
 
 The context type schema for a pattern `p` is:
 
-*   **List binder or matcher**: A type schema `List<E>` where:
-    *   If `p` has a type argument, then `E` is the type argument.
-    *   Else `E` is the greatest lower bound of the type schemas of all element
-        subpatterns.
-
-*   **Map binder or matcher**: A type schema `Map<K, V>` where:
-    *   If `p` has type arguments then `K`, and `V` are those type arguments.
-    *   Else `K` is the least upper bound of the types of all key expressions
-        and `V` is the greatest lower bound of the context type schemas of all
-        value subpatterns.
+*   **Or**: The least upper bound of the context type schemas of the branches.
 
-*   **Record binder or matcher**: A record type scheme with positional and named
-    fields corresponding to the type schemas of the corresponding field
-    subpatterns. If a named field uses the shorthand syntax to infer a variable
-    subpattern with the same name as the field, then the type schema is `?` for
-    that field.
+*   **And**: The greatest lower bound of the context type schemas of the branches.
 
-    *Note that the type schema will be a record type even when the matched value
-    type isn't a record, as in:*
+*   **Null-check** or **null-assert**: A context type schema `E?` where `E` is
+    the context type schema of the inner pattern. *For example:*
 
     ```dart
-    var p = Point(x: 1, y: 2);
-    var (x: x, y: y) = p;
+    var [[int x]!] = [[]]; // Infers List<List<int>?> for the list literal.
```

Comment:
I think of the context type of a pattern more like "the type it could reasonably *attempt* to match" but not necessarily the type it *will* match.

If the user took the time to write a null-check or null-assert pattern, that implies that they *do* expect `null` to sometimes flow into that pattern (and be refuted or throw). Otherwise, they would have simply left off the `!` or `?` entirely.

But, yes, context types on refutable patterns don't often come into play. (Though I guess they do now that I tweaked the proposal to use a context type for pattern-ifs.)

Reply:
That's a good point. You're right that we already say that we don't even calculate a context type for switch cases. So there really is no point in defining it at all for refutable patterns.

However, I just tweaked the proposal to say that we do use a context type for pattern-if statements, so now the context type schema does come into play with refutable patterns, but *only* for those. If there *is* going to be a context type at all for a refutable pattern, I think it may as well be the full one.